docs/rfcs/blocks-in-widget-areas.md

```diff
@@ -0,0 +1,520 @@
+		'recent-posts-2',
+		'recent-comments-2',
+	),
+	'array_version' => 4,
+)
+```
+
+Here, `123` is the ID of a `wp_area` post. The `post_content` field of `get_post( 123 )` contains the `'footer'` widget area's block markup.
+
+```html
+<!-- wp:paragraph --><p>Hello there!</p><!-- /wp:paragraph -->
+<!-- wp:legacy-widget {"identifier":"search-2","instance":{"title":""}} /-->
+<!-- wp:legacy-widget {"identifier":"recent-posts-2","instance":{"title":"","number":5}} /-->
+<!-- wp:legacy-widget {"identifier":"recent-comments-2","instance":{"title":"","number":5}} /-->
+```
+
+## Backwards Compatibility
+
+### `wp_get_sidebars_widgets()`
+
+[`wp_get_sidebars_widgets()`][wp_get_sidebars_widgets] can now potentially return an associative array that maps strings to integers. This differs from the current behaviour which is to always return an associative array that maps strings to arrays.
+
+Since `wp_get_sidebars_widgets()` is marked `@private`, no effort is made to ensure backwards compatibility beyond incrementing the `'array_version'` field from `3` to `4`.
+
+The same goes for [`wp_set_sidebars_widgets()`][wp_set_sidebars_widgets] and `get_option( 'sidebars_widgets' )`.
+
+[wp_get_sidebars_widgets]: https://developer.wordpress.org/reference/functions/wp_get_sidebars_widgets/
+[wp_set_sidebars_widgets]: https://developer.wordpress.org/reference/functions/wp_set_sidebars_widgets/
+
+### `'sidebars_widgets'` filter
+
+The [`'sidebars_widgets'` filter][sidebars_widgets] remains as-is and is passed an associative array that maps widget area identifiers to an an array of widget identifiers.
+
+To do this, sequences of blocks in a widget area will be converted into dynamically registered _blocks widgets_ which outputs the blocks in that sequence.
+
+To illustrate, let's say that the `'footer'` widget area has been migrated.
+
+```php
+array(
+	'footer' => 123,
+	'array_version' => 4,
+)
+```
+
+The referenced `wp_area` contains some blocks and a legacy widget in its `post_content`.
+
+```html
+<!-- wp:paragraph --><p>Hello there!</p><!-- /wp:paragraph -->
+<!-- wp:legacy-widget {"identifier":"search-2","instance":{"title":""}} /-->
+<!-- wp:separator --><hr class="wp-block-separator"/><!-- /wp:separator -->
+<!-- wp:paragraph --><p>Welcome to my cool website!</p><!-- /wp:paragraph -->
+```
+
+Now, let's take a look at what the `'sidebars_widgets'` filter receives.
+
+```php
+function my_sidebars_widgets_filter( $sidebars_widgets ) {
+	var_dump( $sidebars_widgets );
+	return $sidebars_widgets;
+}
+add_filter( 'sidebars_widgets', 'my_sidebars_widgets_filter' );
+```
+
+We receive an array that contains three widgets.
+
+```php
+array(
+	'footer' => array(
+		'wp-blocks-cb8d16f47bf5cfa6ef313e813d6b80d6',
+		'search-2',
+		'wp-blocks-d93dcd8e370f526bf072efcb5138dc53',
+	),
+	'array_version' => 3,
+)
+```
+
+The first `wp-blocks-{hash}` widget is a dynamically registered widget that outputs a paragraph, and the second outputs a separator and a paragraph.
```

Comment:
Why are the last two merged into one:
```html
<!-- wp:paragraph --><p>Hello there!</p><!-- /wp:paragraph -->
<!-- wp:legacy-widget {"identifier":"search-2","instance":{"title":""}} /-->
<!-- wp:separator --><hr class="wp-block-separator"/><!-- /wp:separator -->
<!-- wp:paragraph --><p>Welcome to my cool website!</p><!-- /wp:paragraph -->
```
results in:
```php
array(
	'footer' => array(
		'wp-blocks-cb8d16f47bf5cfa6ef313e813d6b80d6',
		'search-2',
		'wp-blocks-d93dcd8e370f526bf072efcb5138dc53',
	),
	'array_version' => 3,
)
```

Reply:
I attempted to explain why in the paragraph below. Does it make sense?

> This is done so that all blocks aren't wrapped with `'before_widget'` and `'after_widget'` markup. Instead, **sequences** of blocks are combined together _and then_ wrapped with `'before_widget'` and `'after_widget'` markup. This makes blocks render on old themes as if they were all together in a single widget.